File: BookLand_BE/src/main/java/com/example/bookland_be/entity/Event.java
package com.example.bookland_be.entity;

import lombok.*;
import jakarta.persistence.*;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "event")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Event {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String name;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(nullable = false)
    private String type;

    @Column(nullable = false)
    private LocalDateTime startTime;

    @Column(nullable = false)
    private LocalDateTime endTime;

    @Enumerated(EnumType.STRING)
    @Builder.Default
    private EventStatus status = EventStatus.DRAFT;

    @Builder.Default
    private Integer priority = 0;

    @Column(updatable = false)
    private LocalDateTime createdAt;

    @OneToMany(mappedBy = "event", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private Set<EventTarget> targets = new HashSet<>();

    @OneToMany(mappedBy = "event", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private Set<EventRule> rules = new HashSet<>();

    @OneToMany(mappedBy = "event", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private Set<EventAction> actions = new HashSet<>();

    @OneToMany(mappedBy = "event", cascade = CascadeType.ALL)
    @Builder.Default
    private Set<EventLog> logs = new HashSet<>();

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
    }

    // Helper method to check if event is active
    @Transient
    public boolean isActive() {
        LocalDateTime now = LocalDateTime.now();
        return status == EventStatus.ACTIVE
                && now.isAfter(startTime)
                && now.isBefore(endTime);
    }

    public enum EventStatus {
        DRAFT, ACTIVE, PAUSED, EXPIRED, DISABLED
    }
}

File: BookLand_FE/src/types/Event.ts
import type { User } from './User';
import type { EventImage } from './EventImage';
import type { EventTarget } from './EventTarget';
import type { EventRule } from './EventRule';
import type { EventAction } from './EventAction';
import type { EventLog } from './EventLog';
import type { EventType } from './EventType';

export const EventStatus = {
    DRAFT: 'DRAFT',
    ACTIVE: 'ACTIVE',
    PAUSED: 'PAUSED',
    EXPIRED: 'EXPIRED',
    DISABLED: 'DISABLED'
} as const;

export type EventStatus = (typeof EventStatus)[keyof typeof EventStatus];

export interface Event {
    id: number;
    name: string;
    description?: string;
    type: EventType;
    startTime: string;
    endTime: string;
    status: EventStatus;
    priority: number;
    createdBy?: User;
    createdAt?: string;
    images?: EventImage[];
    targets?: EventTarget[];
    rules?: EventRule[];
    actions?: EventAction[];
    logs?: EventLog[];
}
